File: leetcode/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cc
// https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
// Medium
// June 8th, 2021 Challenge

#include<vector>
#include<unordered_set>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    unordered_set<int> s;
    int pre = 0, in = 0;
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (in >= preorder.size() || pre >= preorder.size()) return nullptr;
        if (s.find(inorder[in]) != s.end()) return nullptr;
        TreeNode* curr = new TreeNode(preorder[pre]);
        s.insert(preorder[pre]);
        if (preorder[pre] == inorder[in]) {
            ++pre;
            ++in;
            curr->right = buildTree(preorder, inorder);
            return curr;
        }
        pre++;
        curr->left = buildTree(preorder, inorder);
        in++;
        curr->right = buildTree(preorder, inorder);
        return curr;
    }
};

File: leetcode/python/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
# Medium

# Key is to find preorder[0] in inorder. Then whatever index it exists on inorder, the left
#   of that is the left of the tree, and the right of that is the right of the tree
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        ret = TreeNode(preorder[0])
        
        i = inorder.index(preorder[0])
        if i > 0:
            ret.left = self.buildTree(preorder[1:i+1], inorder[:i])
        if i < len(inorder)-1:
            ret.right = self.buildTree(preorder[i+1:], inorder[i+1:])
        return ret